resource_work_time_from_contracts/models/resource_mixin.py

```diff
@@ -104,56 +35,209 @@ def get_work_days_data(
+        # first one found will be used.
+        current_contracts = self._get_active_contracts(date, date)
+        if not current_contracts:
+            return None
+        return current_contracts[0].resource_calendar_id
+
+    def _sum_intervals(self, intervals):
+        result = defaultdict(float)
+        for start, stop, meta in intervals:
+            result[start.date()] += (stop - start).total_seconds() / 3600
+        return sorted(result.items())
+
+    def _get_round_day_round_func(self, rounding_mode, granularity):
+        def round_day_default(work_time, normal_work_time):
+            # this is the same rounding computation as in
+            # resource.resource_mixin.get_work_days_data().
+            return (
+                float_utils.round(ROUNDING_FACTOR * work_time / normal_work_time)
+                / ROUNDING_FACTOR
+            )
+
+        def round_day_round(work_time, normal_work_time):
+            return (
+                float_utils.round((work_time / normal_work_time) / granularity)
+                * granularity
+            )
+
+        def round_day_ceil(work_time, normal_work_time):
+            return (
+                math.ceil(
+                    (work_time / normal_work_time) / granularity
+                    - granularity * DAY_ROUNDING_CEIL_THRESHOLD
+                )
+                * granularity
+            )
+
+        if rounding_mode == DAY_ROUNDING_MODE_ROUND:
+            return round_day_round
+        if rounding_mode == DAY_ROUNDING_MODE_CEIL:
+            return round_day_ceil
+        return round_day_default
+
+    def _get_day_rounding_func(self):
+        ir_config_parameter_model = self.env["ir.config_parameter"].sudo()
+        rounding_mode = ir_config_parameter_model.get_param(DAY_ROUNDING_MODE_PARAM)
+        rounding_granularity = ir_config_parameter_model.get_param(
+            DAY_ROUNDING_GRANULARITY_PARAM
+        )
+        if rounding_granularity:
+            rounding_granularity = float(rounding_granularity)
+        else:
+            rounding_granularity = DEFAULT_DAY_ROUNDING_GRANULARITY
+        return self._get_round_day_round_func(rounding_mode, rounding_granularity)
+
+    def _sum_work_days_data(self, work_time_per_day, normal_work_time_per_day):
         num_days = 0.0
         num_hours = 0.0
+        round_day = self._get_day_rounding_func()
         for day, work_time in work_time_per_day:
             if work_time == 0.0:
                 continue
             normal_work_time = normal_work_time_per_day[day]
-            # we use the same rounding computation as in
-            # resource.resource_mixin.get_work_days_data().
-            num_days += (
-                float_utils.round(ROUNDING_FACTOR * work_time / normal_work_time)
-                / ROUNDING_FACTOR
-            )
+            num_days += round_day(work_time, normal_work_time)
             num_hours += work_time
```

Comment:
Are we certain that this should be continuously rounded in the loop, instead of rounded once at the end?

Reply:
yes. this is how it’s done in the `resource` module too. it’s the value for each day that must be rounded on its own, and then all of these summed together. this is especially true when using the non-default rounding methods, where the value of each day is rounded according to a defined granularity.